site/public/js/rainbow-customization.js

```diff
@@ -267,78 +267,144 @@ function getGradeableBuckets() {
+    const thirdInput = document.getElementById(tableMap[table][3]).value.trim();
+
+    // Check whether input is allowed
     // eslint-disable-next-line no-undef
     const studentFullDataValues = studentFullData.map((item) => item.value);
-    if (!studentFullDataValues.includes(USERID)) {
-        alert('Invalid User ID. Please enter a valid one.');
-        return;
-    }
+    const tableBody = document.getElementById(tableMap[table][0]);
+    const rows = tableBody.getElementsByTagName('tr');
+    switch (table) {
+        case 0: { // plagiarism table
+            if (firstInput === '' || secondInput === '' || thirdInput === '') {
+                alert('Please fill in all the fields.');
+                return;
+            }
+            if (!studentFullDataValues.includes(firstInput)) {
+                alert('Invalid User ID. Please enter a valid one.');
+                return;
+            }
+            if (thirdInput > 1 || thirdInput < 0) {
+                alert('Penalty must be between 0 - 1');
+                return;
+            }
 
-    // Check for penalty
-    if (penalty > 1 || penalty < 0) {
-        alert('Penalty must be between 0 - 1');
-        return;
-    }
+            // Check for duplicate entries
+            for (let i = 0; i < rows.length; i++) {
+                const row = rows[i];
+                const existingFirstInput = row.cells[0].textContent.trim();
+                const existingSecondInput = row.cells[1].textContent.trim();
 
-    const tableBody = document.getElementById('table-body');
+                if (firstInput === existingFirstInput && secondInput === existingSecondInput) {
+                    alert('Entry with the same Student ID and Gradeable already exists.');
+                    return;
+                }
+            }
+            break;
+        }
+        case 1: { // manual grading table
+            if (firstInput === '' || secondInput === '') {
+                alert('Please fill in both user ID and final grade.');
+                return;
+            }
+            if (!studentFullDataValues.includes(firstInput)) {
+                alert('Invalid User ID. Please enter a valid one.');
+                return;
+            }
+            const allowed_grades = ['A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D+', 'D', 'F'];
+            if (!allowed_grades.includes(secondInput)) {
+                alert('Grade must be one of the following: A, A-, B+, B, C+, C, C-, D+, D, F');
+                return;
+            }
 
-    // Check for duplicate entries
-    const rows = tableBody.getElementsByTagName('tr');
-    for (let i = 0; i < rows.length; i++) {
-        const row = rows[i];
-        const existingUSERID = row.cells[0].textContent.trim();
-        const existingGradeable = row.cells[1].textContent.trim();
+            // Check for duplicate entries
+            for (let i = 0; i < rows.length; i++) {
+                const row = rows[i];
```

Comment:
A [`.foreach()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) could be helpful here.

Reply:
There's no way to return from a Javascript `.forEach()`, which makes things a bit needlessly complicated.